### ISSUE TITLE
core - default safety belts on 

### ISSUE THREAD

**Issue by kapilt** [cloud-custodian/cloud-custodian#3641]
i've been talking to a few different long time users, and they all have stories of policies gone awry, mostly due to human error, and over the years we've placed various safety belts in the system to help combat that, like max-resources, and max-resources-percent, which stop a policies execution before it ever modifies if a resource if it would exceed a user defined threshold of resources it would touch. The problem is the safeties default to off, and custodian is a tool of mass operation, rather than teaching folks after the fact, i'm thinking we should change the defaults, and force folks to explicitly opt in on policies that may have mass fleet implications. When this topic has come up before we've always aligned to our compatibility guarantees on policies, so this may well qualify to some as a breaking change. But i feel we have to balance safety with our compatibility constraints better, and err on the side of not letting our users accidentally shoot themselves in the foot.

So what would it look like, basically if a policy doesn't have any actions nothing changes. If it only uses a set of whitelisted actions, it also wouldn't be unaffected. But ifs its potentially causing destructive action across an arbitrary percent.. say 42.. then stop policy execution with a helpful failure message. The threshold is already configurable, so this is simply applying a default non zero value.

**Comment by nitrocode** [comment]
> But i feel we have to balance safety with our compatibility constraints better, and err on the side of not letting our users accidentally shoot themselves in the foot.

I agree with this statement. It's possible to make extremely destructive changes very easily. I've probably run the `--dryrun` command over 100 times for each policy, checking the output directory, before removing that flag as I'm paranoid I'll break something.
